### PR TITLE
refactor(libanki): extract 'Field/Fields' 

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
I suggested this in #17647, 

## Approach
* Create `Fields`
  * Update `getJSONObject("flds") to use this
* Create `Field`
  * Fix compile errors
  * Document and add a test
  
## How Has This Been Tested?
Tested all field related functionality on 'Manage Note Types', API 34 emulator

## Learning

We're missing a number of features, admittedly some of these are Anki Desktop Note Editor only
 * [Anki manual - fields](https://docs.ankiweb.net/templates/fields.html)
 * [Anki manual - customizing fields](https://docs.ankiweb.net/editing.html#customizing-fields)

⚠️ `JSONObject.getString()` appears to differ between `JvmTest` and `AndroidTest` regarding nulls and objects.

This sounds wrong, so please confirm

# Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
boldly assuming related PR makes it through merge queue where it sits now, removing blocked label

**Review by BrayanDSO — APPROVED** [review]
Really neat

File: AnkiDroid/src/main/java/com/ichi2/libanki/Field.kt [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * Basic field substitution syntax looks like: `{{Front}}`, where `Front` is defined by [name].
+ * This syntax is Anki-specific (similar to [mustache](https://mustache.github.io/mustache.5.html]))
+ * and is [documented in the Manual](https://docs.ankiweb.net/templates/fields.html)
+ *
+ * This class documents field properties, most of which are relevant to inserting/editing notes,
+ * with a few related to searching/card rendering
+ *
+ * In AnkiDroid, this also affects the font and language used in 'type the answer'
+ *
+ * ### Undocumented properties:
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Field.kt:37]: I find it strange to only note that they are "undocumented".

I think I'd rewrite this as:
> ### Desktop only properties
>
> In this class, we create helper variables and values only when we use them. For this reason, 
> Desktop (and maybe ankimobile) version use some field that are not (currently) being used in ankidroid.
> This is a list of those properties as of 2024. It may be incomplete when you read it. If you need to use one of those properties, please create a new variable or value to access and/or edit them

File: AnkiDroid/src/main/java/com/ichi2/libanki/Field.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        get() = jsonObject.getString("name")
+        set(value) {
+            jsonObject.put("name", value)
+        }
+
+    /** The 0-based ordinal of the field */
+    val ord: Int
+        get() = jsonObject.getInt("ord")
+
+    /** Whether the note editor remembers the previous value of the field */
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Field.kt:69]: the way the documentation is written, it seems to indicate that if you close and reopen the note editor, it'll remember this value.
I'd rephrase it as
"If false, when a note is created, the note editor's erase the content of this field"

Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Field.kt:69]: the way the documentation is written, it seems to indicate that if you close and reopen the note editor, it'll remember this value.
I'd rephrase it as
"If false, when a note is created, the note editor's erase the content of this field"

File: AnkiDroid/src/main/java/com/ichi2/libanki/Fields.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    override fun iterator() =
+        jsonArray
+            .jsonObjectIterator()
+            .asSequence()
+            .map(::Field)
+            .iterator()
+
+    operator fun get(index: Int) = Field(jsonArray.getJSONObject(index))
+
+    operator fun set(
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Fields.kt:40]: I'd consider stating explicitly to `@see jsonArray.put`, because the behavior of JSONarray is different from the behavior of arrays, and may be unexpected to a contributor.

**Review by Arthur-Milchior — APPROVED** [review]

**Comment by Arthur-Milchior** [comment]
love this improvement.
All those remarks are nits. If you want to ignore, feel free to press Merge.

File: AnkiDroid/src/main/java/com/ichi2/anki/utils/ext/JSONObject.kt [hunk 4]
```diff
@@ -19,8 +19,16 @@ package com.ichi2.anki.utils.ext
 
 import org.json.JSONObject
 
+/**
+ * @return `null` if:
+ * * The key does not exist
+ * * The value is [null][JSONObject.NULL]
+ * * ⚠️ JVM only. The value is not a string (`{ }` etc...)
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/utils/ext/JSONObject.kt:26]: Thanks so much for the patch!
